### PR TITLE
Staking new era and slash validators fixing

### PR DESCRIPTION
**Pull request by HackFisher**
Not finished yet.

- new_era
- slash_validator
- CurrencyToVote
- remove duplicated phragmen

### CODE REVIEW

File: srml/staking/src/lib.rs [hunk 1]
```diff
@@ -1204,66 +1364,36 @@ impl<T: Trait> Module<T> {
...
-		} else {
-			return;
-		};
-
-		// Early exit if validator is invulnerable.
-		if Self::invulnerables().contains(&stash) {
-			return;
+	/// This is called :
+	/// - Immediately when an account's balance falls below existential deposit.
+	/// - after a `withdraw_unbond()` call that frees all of a stash's bonded balance.
```
Comment (AurevoirXavier) [inline srml/staking/src/lib.rs]: `withdraw_unbond` was removed in https://github.com/darwinia-network/darwinia/pull/92. Reserve?

File: srml/staking/src/lib.rs [hunk 2]
```diff
@@ -66,6 +68,32 @@ const STAKING_ID: LockIdentifier = *b"staking ";
...
+	total: Points,
+	/// The reward points earned by a given validator. The index of this vec corresponds to the
+	/// index into the current validator set.
+	individual: Vec<Points>,
+}
+
+impl EraPoints {
+	/// Add the reward to the validator at the given index. Index must be valid
+	/// (i.e. `index < current_elected.len()`).
+	fn add_points_to_index(&mut self, index: u32, points: u32) {
```
Comment (AurevoirXavier) [inline srml/staking/src/lib.rs]: `points: Points` looks better.

File: srml/staking/src/lib.rs (hunk 1 above)
Comment (HackFisher) [reply srml/staking/src/lib.rs]: No need to reserve, if #92 is merged first, this PR will do the merging and resolve conflicts.

**Comment by AurevoirXavier** [comment]
`cargo build` in srml/staking

```rust
error: cannot find macro `vec` in this scope
   --> /Users/xavier/.cargo/git/checkouts/substrate-a7fa553ead635512/d2c4b0d/core/phragmen/src/lib.rs:272:40
    |
272 |         let mut assignment = (n.who.clone(), vec![]);
    |                                              ^^^

error: aborting due to previous error

error: could not compile `substrate-phragmen`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Pull request merged**